modules/ximgproc/src/globalmatting.cpp

```diff
@@ -0,0 +1,568 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+#include "precomp.hpp"
+
+
+namespace cv
+{
+namespace ximgproc
+{
+template <typename T>
+inline T sqr(T a)
+{
+           return a * a;
+}
+
+struct IntensityComp
+{
+    IntensityComp(const cv::Mat_<cv::Vec3b> &img_temp) : img(img_temp)
+    {
+
+    }
+
+    bool operator()(const cv::Point &p0, const cv::Point &p1) const
+    {
+	const cv::Vec3b &c0 = img(p0.y, p0.x);
+	const cv::Vec3b &c1 = img(p1.y, p1.x);
+
+	return ((int)c0[0] + (int)c0[1] + (int)c0[2]) < ((int)c1[0] + (int)c1[1] + (int)c1[2]);
+    }
+
+    const cv::Mat_<cv::Vec3b> &img;
+};
+
+
+
+
+
+std::vector<cv::Point> GlobalMatting::findBoundaryPixels(const cv::Mat_<uchar> &trimap, int a, int b)
+{
+    std::vector<cv::Point> result;
+
+    for (int x = 1; x < trimap.cols - 1; ++x)
+	for (int y = 1; y < trimap.rows - 1; ++y)
+	{
+	    if (trimap(y, x) == a)
+	    {
+	        if (trimap(y - 1, x) == b ||
+	            trimap(y + 1, x) == b ||
+	            trimap(y, x - 1) == b ||
+	            trimap(y, x + 1) == b)
+	        {
+	            result.push_back(cv::Point(x, y));
+	        }
+	    }
+	}
+
+    return result;
+}
+
+// Eq. 2
+float GlobalMatting::calculateAlpha(const cv::Vec3b &F, const cv::Vec3b &B, const cv::Vec3b &I)
+{
+    float result = 0;
+    float div = 1e-6f;
+    for (int c = 0; c < 3; ++c)
+    {
+	float f = F[c];
+	float b = B[c];
+	float i = I[c];
+
+	result += (i - b) * (f - b);
+	div += (f - b) * (f - b);
+    }
+
+    return std::min(std::max(result / div, 0.f), 1.f);
+}
+
+// Eq. 3
+float GlobalMatting::colorCost(const cv::Vec3b &F, const cv::Vec3b &B, const cv::Vec3b &I, float alpha)
+{
+    float result = 0;
+    for (int c = 0; c < 3; ++c)
+    {
+	float f = F[c];
+	float b = B[c];
+	float i = I[c];
+
+	result += sqr(i - (alpha * f + (1 - alpha) * b));
+    }
+
+    return sqrt(result);
+}
+
+// Eq. 4
+float GlobalMatting::distCost(const cv::Point &p0, const cv::Point &p1, float minDist)
+{
+    int dist = sqr(p0.x - p1.x) + sqr(p0.y - p1.y);
```

Comment:
`normL2Sqr<int>(p0 - p1);`

Reply:
@alalek , I couldn't find documentation of normL2Sqr anywhere

Reply:
[Code is here](https://github.com/opencv/opencv/blob/4.1.1/modules/core/include/opencv2/core/types.hpp#L1414).